### ISSUE TITLE
Feature Request - Compiler warning for calling functions in another bank

### ISSUE THREAD

**Issue by huderlem** [rednex/rgbds#202]
**Current Problem:**  I'm refactoring a large chunk of code and splitting things out into different ROM Banks.  This leaves behind many function calls that assume the function is located within the same Bank as the caller.  The code compiles successfully, but things will break at runtime, since the calls are now spanning ROM Banks.

**Potential Solution:**  Add an optional compiler warning that detects when a `call` is performed on a label that is not located in the same Bank as the caller.  This would disregard any calls to a label in Bank 0 or RAM, since those are always accessible by a caller.  It would also be useful if there were a compiler directive you could use to suppress the warning if needed.

My understanding is that this would not be a trivial feature to implement since the linker might not currently have enough information to make that decision.

**Comment by AntonioND** [comment]
Well, I actually thought about doing that (I even wrote a python script some time ago to do this check by analyzing the code) and started to prototype a patch to add this functionality. The main issue is that it still isn't enough to prevent problems with ROM switching. You could be loading a label of a bank that isn't the current one, for example, and calling a function in ROM0 that is going to do something with it. But you could also be trying to read from it directly, which isn't valid. Or you could have N RAM banks with the same struct and just use the label of the first one to refer to all of them and use them as an array of banks. It's a bit of a mess...

**Comment by huderlem** [comment]
Sure, there are certainly edge cases.  The most common case, though, is simply:
 `call FunctionInDifferentBank`

If it's an optional flag you can turn on for debugging, I believe that would certainly be worth it.

**Comment by aaaaaa123456789** [comment]
This is one of those things that keeps coming up (I don't remember an issue being made explicitly for this purpose, but it certainly has come up in conversation more than once), mostly because, for the "nice" code, it seems like a trivial improvement that should have always been there.

Let me elaborate a bit on that. The "nice" code is essentlally what everyone learns to write. It's what you'll find in examples everywhere, what is considered clean code; beginners will always write this way, and non-beginners will continue to write this way 99% of the time because it's simply a lot easier to write. In the case of function calls, the "nice" way is to always call a label at the top of a function: `call MyFunction`.

And then you have the ugly code. The ugly code will happily `call $4000` because it knows that every bank that could be loaded at the time will have a valid function there, even if not the same one. The ugly code will `call wRelocatedCodeBase + (MyFunction - RelocatableCodeStart)` after copying a bunch of functions to WRAM. The ugly code will call a function in the wrong bank, simply because the programmer knows that the function has been replicated in multiple banks and doesn't feel like defining and remembering multiple labels for what is essentially the same function. 
All of these tricks, and probably quite a few more I'm simply not remembering right now, are found in real codebases. 99% of the code doesn't use them; code becomes pretty unreadable when it's full of them. But they have their place, and they have been used since time immemorial.

An assembler that intends to be generally useful (as opposed to, say, an assembler to be used only as the final pass of a specific compiler) should support these forms, and rgbasm certainly does — the argument to `call` is an expression, not necessarily a 16-bit label. But support for these convoluted cases simply muddies up any attempt to give warnings for functions that aren't accessible from the current bank.

It should also be noted that functions are only part of the problem, as accessing data outside of the bank where it exists suffers from the same issues. There's also the issue of accessing banked objects from ROM0, as you have to ensure that the correct bank is loaded, but there is no way for this or any toolchain to verify that this happens, short of executing the code up to the point where the questioned instruction lies.

There aren't many courses of action that I can think of that would alleviate this issue without breaking the "ugly" code. Besides the null course of action ("do nothing"), a few ideas I can think of follow, but all of them come with some sort of downside.

**1) Checked labels**

A new pseudo-instruction, let's call it `checkbank`, would have to be added. This instruction would accept a label (or optionally, a list of labels) as its argument, and verify if the label(s) passed to it belong to the same ROM bank as the code that invokes it. The assembler would emit no bytes for this instruction, but rather some sort of linker directive; the linker would simply verify that the banks match, and fail to link if they don't. 
Thus, a call that checks whether its target is on the same ROM bank would look like:

```
  checkbank MyFunction
  call MyFunction
```

Note that the complexity of needing two very repetitive lines to call a function can be absorbed into a macro. Also, for this to work, `checkbank` would have to succeed for labels in ROM0, which seems straightforward anyway.

Also note that this would work for data as well, not just for functions.

Downsides:

* This requires defining a new pseudo-instruction, that might collide with user-defined macros in other codebases.
* As it is an opt-in check, it wouldn't work for existing code unless all of it is modified to use this check.
* Since the only thing that can be done at the linker stage is fail, this is useful when functions call other functions in other banks, but not when functions incorrectly farcall functions in the same bank (which is an optimization opportunity that should be detected). This can be helped by adding another pseudo-instruction that does the opposite of this one.

**2) Automatic label checking**

Instead of defining a new pseudo-instruction, incorporate the `checkbank` semantics explained above into `call` whenever the argument is a label. This amounts to pretty much what OP is requesting.

Downsides:

* It requires detecting whether the argument to `call` is a single label or a more complex expression. This would probably require some serious lexer changes.
* `call MyFunction` and `call MyFunction + 0` would have different semantics. This is an egregious violation of the principle of least surprise, as it would confuse pretty much anyone encountering that such an instruction for the first time. The issue is compounded by the fact that nobody would expect `call MyFunction + 0` to be at all useful.
* This new behavior would be backwards-incompatible, and thus would probably hold back many users from updating their toolchain.
* If the `+ 0` loophole was closed, this would fall back into an issue similar to the ld/ldh one (not having any way of specifying `ld [$ff00], a` without it being optimized), only that it would effectively prevent whole codebases from building at all. Requiring codebases to fall back to `dbw $cd, MyFunction` just to be able to compile is a bad idea. (`dbw` is a pokecrystal macro, but its purpose should be obvious.)
* This still suffers from not being able to detect farcalls into the same bank, but without a way to solve this particular issue.
* While an assembler flag can help with some of the issues above, it is still an "all or nothing" toggle, and thus not particularly useful; its only real purpose would be to hide this behavior from entire codebases (and avoid backwards incompatibility).

**3) Bank-checking function**

While it is possible to write `if BANK(MyFunction) == BANK(@)`, this is not very useful, as the assembler will often (if not always, you tell me) be unable to evaluate that expression at all. Therefore, a function `BANKCHECK(MyFunction)` could return 1 if the given label is accessible from the current ROM bank (i.e., it is in ROM0 or in the current bank) or 0 otherwise. 
There are two possible ways of implementing this:

* The "pessimistic" way, which just results in 0 whenever it can't determine whether the label would be in the current bank or not;
* The "accurate" way, which would defer the evaluation of this expression until it can determine this accurately.

Downsides:

* The pessimistic way would simply result in a lot of 0s. The assembler cannot usually determine this, and if it can, the programmer can determine it easily as well.
* The accurate way requires deferring calculation to the linker. This completely disallows using it in conditional call/farcall macros, since the argument to an `if` cannot be evaluated by the linker.
* The only purpose this function could ever serve, then, would be to generate a macro-based implementation of the `checkbank` pseudo-instruction described above, which comes with its own downsides.

---

All in all, there isn't much that can be done toolchain-side to avoid the effects of banks. An opt-in per-call check might help, but that's as far as the tools can really go.

**Comment by yenatch** [comment]
Rather than detect when something is being called, just warn about any label from another bank that doesn't also come with `BANK(label)` or `BANK("its section")` in the same scope. Any label that doesn't follow this ought to have an assert anyway. If the rest of your codebase doesn't work this way, you could turn this warning on while refactoring and turn it off when you're done. It should be off by default since any project with more than one bank will have warnings.

I was looking for the feature request for asserts but it looks like @aaaaaa123456789 hasn't made it yet. That would probably be a prerequisite.

**Comment by aaaaaa123456789** [comment]
@yenatch, I created the feature request.

Anyway, while that flag may be useful for some codebases, it's extremely easy to think of cases where no assertion makes sense, yet the labels don't belong to the current bank. This file comes to mind: https://github.com/pret/pokecrystal/tree/master/data/battle/effect_command_pointers.asm

This is the very reason why I strongly disagree with anything but a per-instance opt-in check. This is yet another example of nice code vs. cool tricks that I elaborated on in my previous message. Any attempt at addressing it (such as "`dw` doesn't count") results in inconsistent behavior that will cause problems in actual codebases. At the worst end of this spectrum we have handling lone labels as a special case, since that causes `SomeLabel` and `SomeLabel + 0` to have different semantics; however, there are just too many cases that can't be caught by a blanket rule like this one. 1% is a lot in a million lines of code.

All in all, the only reasonable approach is to add bank verification _to the lines of code that need it_, which can easily be done with a macro. Link-time assertions would allow implementing a `checkbank` macro that behaved like the one I described above, which could in turn be used to implement checked calls. That should be enough for most codebases.

**Comment by yenatch** [comment]
What I meant to say was that asserts would somehow cover labels that would otherwise cause a warning. I'm not sure how that would be done. Maybe any assert that includes that label (or maybe `BANK(label)` in this case) would suffice, with the drawback that it might not always warn when it should.

Using that table as an example, you could make a macro that asserted each label was in the right bank. This is exactly the sort of thing that should be asserted.

```asm
battle_command_pointer: MACRO
	assert BANK(\1) == BANK("Effect Commands")
	dw \1
ENDM

BattleCommandPointers:
; entries correspond to macros/scripts/battle_commands.asm
	battle_command_pointer BattleCommand_CheckTurn
	battle_command_pointer BattleCommand_CheckObedience
	battle_command_pointer BattleCommand_UsedMoveText
	battle_command_pointer BattleCommand_DoTurn
	; etc
```

Assuming that the point of this flag would be for refactoring, I don't see a problem with complaining about code that can't reasonably be asserted. Again, if you really wanted to leave it on permanently, you could suppress the warning for individual cases with something silly like `assert BANK(Label)`. If the majority of your code is like this you're not likely to get much use out of this feature anyway.

Also, since this would be an rgbasm flag, this could be done per-object. Though messy codebases are more likely to have only one object.

**Comment by aaaaaa123456789** [comment]
>Assuming that the point of this flag would be for refactoring, I don't see a problem with complaining about code that can't reasonably be asserted, especially if it's temporary.

Makes sense in that context. I guess it's not a problem if the flag isn't meant for general, everyday use.

**Comment by ISSOtm** [comment]
My current view on this issue is that you could, assuming #488 gets merged, write the following macro:
```
checked_call: MACRO
    call \1
    assert FAIL, BANK(@) == BANK(\1)
ENDM
```
and it would do the job just fine. I don't think compiler support is necessary beyond assertions; if that's agreed on, then I'll consider #488 to close this.

----
A more complete macro would be this:
```
checked: MACRO
    \1, \2
    assert FAIL, BANK(@) == BANK(\2)
ENDM
```
but that's besides the point.

**Comment by ISSOtm** [comment]
Closed by #488.
